File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    maxHealth = health
    curTime = attackIdx = healTime = 0 # 현재 초
    lastTime = attacks[-1][0] # 공격을 받는 마지막 시간
    spendTime = bandage[0]
    healPerSec = bandage[1]
    addHeal = bandage[2]
    while curTime <= lastTime:
        # print('지금 시간은', curTime, health, attacks[attackIdx], healTime)
        if curTime == attacks[attackIdx][0]: # 공격 받을 시간
            health -= attacks[attackIdx][1]
            attackIdx += 1
            healTime = 1
        else:
            addHealAmount = 0
            if healTime == spendTime: #붕대 다 감음
                addHealAmount = healPerSec + addHeal
                healTime = 1
            else:
                addHealAmount = healPerSec
                healTime += 1
            health = min(maxHealth, health + addHealAmount) # 힐링~
        # print(curTime, health)
        if health <= 0: # 죽을 경우
            # print('사망!')
            return -1
        curTime += 1
    return health

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function solution(bandage, health, attacks) {
    let maxHealth = health;
    let curTime = 0;
    let attackIdx = 0;
    let healTime = 0;
    let lastTime = attacks[attacks.length - 1][0];
    let spendTime = bandage[0];
    let healPerSec = bandage[1];
    let addHeal = bandage[2];

    while (curTime <= lastTime) {
        if (curTime === attacks[attackIdx][0]) {
            // 공격 받음
            health -= attacks[attackIdx][1];
            attackIdx += 1;
            healTime = 1;
        } else {
            let addHealth = healPerSec;
            if (healTime === spendTime) {
                // 붕대 다 감은 경우에만 추가적으로 체력 세이브
                addHealth += addHeal;
                healTime = 1;
            } else {
                healTime += 1;
            }
            health = Math.min(addHealth + health, maxHealth);
        }
        console.log(curTime, health);
        if (health <= 0) {
            return -1; // 죽은 경우
        }
        curTime += 1;
    }
    return health;
}
